packages/router/router.ts

```diff
@@ -3038,12 +3087,129 @@ function shouldRevalidateLoader(
   return arg.defaultShouldRevalidate;
 }
 
+async function callRoutePipeline(
+  request: Request,
+  matches: AgnosticDataRouteMatch[],
+  requestContext: unknown,
+  handler:
+    | LoaderFunction
+    | ActionFunction
+    | LoaderFunctionWithMiddleware
+    | ActionFunctionWithMiddleware
+) {
+  // Avoid memory leaks since we don't control the key
+  let store = new WeakMap();
+  let middlewareContext: MiddlewareContext = {
+    get<T>(k: MiddlewareContextInstance<T>) {
+      if (store.has(k)) {
+        return store.get(k) as T;
+      }
+      return k.getDefaultValue();
+    },
+    set<T>(k: MiddlewareContextInstance<T>, v: T) {
+      if (typeof v === "undefined") {
+        throw new Error(
+          "You cannot set an undefined value in the middleware context"
+        );
+      }
+      store.set(k, v);
+    },
+    next: () => {},
+  };
+
+  return callRouteSubPipeline(
+    request,
+    matches,
+    matches[0].params,
+    requestContext,
+    middlewareContext,
+    handler
+  );
+}
+
+async function callRouteSubPipeline(
+  request: Request,
+  matches: AgnosticDataRouteMatch[],
+  params: Params<string>,
+  requestContext: unknown,
+  middlewareContext: MiddlewareContext,
+  handler:
+    | LoaderFunction
+    | ActionFunction
+    | LoaderFunctionWithMiddleware
+    | ActionFunctionWithMiddleware
+): Promise<ReturnType<LoaderFunction>> {
+  if (request.signal.aborted) {
+    throw new Error("Request aborted");
+  }
+
+  if (matches.length === 0) {
+    // We reached the end of our middlewares, call the handler
+    middlewareContext.next = () => {
+      throw new Error(
+        "You may only call `next()` once per middleware and you may not call " +
+          "it in an action or loader"
+      );
+    };
+    return handler({
+      request,
+      params,
+      context: middlewareContext,
+      requestContext,
+    });
+  }
+
+  // We've still got matches, continue on the middleware train.  The `next()`
+  // function will "bubble" back up the middlewares after handlers have executed
+  let nextCalled = false;
+  let next: MiddlewareContext["next"] = () => {
+    nextCalled = true;
+    return callRouteSubPipeline(
+      request,
+      matches.slice(1),
+      params,
+      requestContext,
+      middlewareContext,
+      handler
+    );
+  };
+
+  if (!matches[0].route.middleware) {
+    return next();
+  }
+
+  middlewareContext.next = next;
+  let res = await matches[0].route.middleware({
+    request,
+    params,
+    context: middlewareContext,
+    requestContext,
+  });
+
+  if (nextCalled) {
+    return res;
+  } else {
+    return next();
+  }
```

Comment:
If the middleware resolves without calling `next()` - just keep going down the chain